File: Engine/Crimson/Graphics/Private/RenderContext.cpp
#include "RenderContext.h"

#include "SDLUtils.h"

#include "Core/Logger.h"
#include "Core/BitUtils.h"
#include "Core/File.h"
#include "Core/Path.h"

#include <cstring>
#include <SDL3_shadercross/SDL_shadercross.h>

namespace cge::Private
{
    RenderContext::RenderContext(SDL_Window* window) : Window(window)
    {
        SDL_PropertiesID props = SDL_CreateProperties();

#if defined(CGE_PLATFORM_WINDOWS) // use d3d12 on windows
        SDL_SetBooleanProperty(props, SDL_PROP_GPU_DEVICE_CREATE_SHADERS_DXIL_BOOLEAN, true);
#elif defined(CGE_PLATFORM_APPLE) // use metal on apple platforms
        SDL_SetBooleanProperty(props, SDL_PROP_GPU_DEVICE_CREATE_SHADERS_MSL_BOOLEAN, true);
#else // use vulkan on everything else
        SDL_SetBooleanProperty(props, SDL_PROP_GPU_DEVICE_CREATE_SHADERS_SPIRV_BOOLEAN, true);
#endif

#ifndef NDEBUG
        SDL_SetBooleanProperty(props, SDL_PROP_GPU_DEVICE_CREATE_DEBUGMODE_BOOLEAN, true);
#endif

        CGE_TRACE("Creating device.");
        Device = SDL_CreateGPUDeviceWithProperties(props);
        CGE_SDL_CHECK(Device, "Create device");
        SDL_DestroyProperties(props);

        SDL_PropertiesID deviceProps = SDL_GetGPUDeviceProperties(Device);
        CGE_INFO("Backend: {}", SDL_GetGPUDeviceDriver(Device));
        CGE_INFO("Device: {}", SDL_GetStringProperty(deviceProps, SDL_PROP_GPU_DEVICE_NAME_STRING, "unknown"));
        CGE_INFO("Driver: {}", SDL_GetStringProperty(deviceProps, SDL_PROP_GPU_DEVICE_DRIVER_INFO_STRING, "unknown"));
        SDL_DestroyProperties(deviceProps);

        CGE_TRACE("Associating device with window.");
        CGE_SDL_CHECK(SDL_ClaimWindowForGPUDevice(Device, Window), "Claim window for device");

        SDL_ShaderCross_Init();

        _transferBufferSize = TransferBufferInitialSize;
        _transferBufferOffset = 0;
        _transferBuffer = CreateTransferBuffer(SDL_GPU_TRANSFERBUFFERUSAGE_UPLOAD, _transferBufferSize);
    }

    RenderContext::~RenderContext()
    {
        SDL_ReleaseGPUTransferBuffer(Device, _transferBuffer);
        SDL_ReleaseWindowFromGPUDevice(Device, Window);
        SDL_DestroyGPUDevice(Device);
    }

    SDL_GPUShader* RenderContext::CreateShader(ShaderStage stage, const std::string& name, const std::string& entryPoint, const ShaderInfo& info)
    {
        /*SDL_GPUShaderFormat format = SDL_GetGPUShaderFormats(Device);
        if ((format & SDL_GPU_SHADERFORMAT_SPIRV) != 0)
            format = SDL_GPU_SHADERFORMAT_SPIRV;
        else if ((format & SDL_GPU_SHADERFORMAT_DXIL) != 0)
            format = SDL_GPU_SHADERFORMAT_DXIL;
        else if ((format & SDL_GPU_SHADERFORMAT_MSL) != 0)
            format = SDL_GPU_SHADERFORMAT_MSL;

        std::string fileExtension;
        switch (format)
        {
            case SDL_GPU_SHADERFORMAT_SPIRV:
                fileExtension = "spv";
                break;
            case SDL_GPU_SHADERFORMAT_DXIL:
                fileExtension = "dxil";
                break;
            case SDL_GPU_SHADERFORMAT_MSL:
                fileExtension = "metal";
                break;
            default:
                CGE_FATAL("Unrecognized shader format {}", format);
        }

        auto fullPath = Path::Combine(CGE_CONTENT_DIR, "Shaders", std::format("{}.{}", name, fileExtension));
        auto data = File::ReadBytes(fullPath);

        SDL_GPUShaderStage sdlStage;
        switch (stage)
        {
            case ShaderStage::Vertex:
                sdlStage = SDL_GPU_SHADERSTAGE_VERTEX;
                break;
            case ShaderStage::Pixel:
                sdlStage = SDL_GPU_SHADERSTAGE_FRAGMENT;
                break;
        }

        SDL_GPUShaderCreateInfo shaderInfo
        {
            .code_size = data.size(),
            .code = data.data(),
            .entrypoint = entryPoint.c_str(),
            .format = format,
            .stage = sdlStage,
            .num_samplers = info.NumSamplers,
            .num_storage_textures = info.NumStorageTextures,
            .num_storage_buffers = info.NumStorageBuffers,
            .num_uniform_buffers = info.NumUniforms
        };

        CGE_TRACE("Creating shader.");
        SDL_GPUShader* shader = SDL_CreateGPUShader(Device, &shaderInfo);
        CGE_SDL_CHECK(shader, "Create shader");

        return shader;*/

        SDL_ShaderCross_ShaderStage shaderStage;
        std::string tag;
        switch (stage)
        {
            case ShaderStage::Vertex:
                shaderStage = SDL_SHADERCROSS_SHADERSTAGE_VERTEX;
                tag = "v";
                break;
            case ShaderStage::Pixel:
                shaderStage = SDL_SHADERCROSS_SHADERSTAGE_FRAGMENT;
                tag = "p";
                break;
        }

        auto fullPath = Path::Combine(CGE_CONTENT_DIR, "Shaders", std::format("{}_{}.spv", name, tag));
        auto data = File::ReadBytes(fullPath);

        SDL_ShaderCross_SPIRV_Info spirvInfo
        {
            .bytecode = data.data(),
            .bytecode_size = data.size(),
            .entrypoint = entryPoint.c_str(),
            .shader_stage = shaderStage,
        };

        SDL_ShaderCross_GraphicsShaderResourceInfo resources
        {
            .num_samplers = info.NumSamplers,
            .num_storage_textures = info.NumStorageTextures,
            .num_storage_buffers = info.NumStorageBuffers,
            .num_uniform_buffers = info.NumUniforms
        };

        CGE_TRACE("Compiling shader.");
        SDL_GPUShader* shader = SDL_ShaderCross_CompileGraphicsShaderFromSPIRV(Device, &spirvInfo, &resources, 0);
        CGE_SDL_CHECK(shader, "Compile shader from SPIR-V");
        return shader;
    }

    SDL_GPUTransferBuffer* RenderContext::CreateTransferBuffer(SDL_GPUTransferBufferUsage usage, u32 size) const
    {
        SDL_GPUTransferBufferCreateInfo bufferInfo
        {
            .usage = usage,
            .size = size
        };

        CGE_TRACE("Creating {}KiB transfer buffer.", size / 1024);
        SDL_GPUTransferBuffer* buffer = SDL_CreateGPUTransferBuffer(Device, &bufferInfo);
        CGE_SDL_CHECK(buffer, "Create transfer buffer");

        return buffer;
    }

    SDL_GPUBuffer* RenderContext::CreateBuffer(SDL_GPUBufferUsageFlags usage, u32 size) const
    {
        SDL_GPUBufferCreateInfo bufferInfo
        {
            .usage = usage,
            .size = size
        };

        CGE_TRACE("Creating {}KiB buffer", size / 1024);
        SDL_GPUBuffer* buffer = SDL_CreateGPUBuffer(Device, &bufferInfo);
        CGE_SDL_CHECK(buffer, "Create buffer");

        return buffer;
    }

    SDL_GPUTransferBuffer* RenderContext::GetUploadBuffer(u32 size, u32& offset, bool& shouldCycle)
    {
        if (size >= _transferBufferSize)
        {
            CGE_DEBUG("Requested size ({}KiB) is larger than the current transfer buffer ({}KiB)! A new one will be created.", size / 1024, _transferBufferSize / 1024);
            _transferBufferSize = BitUtils::RoundToNearestPowerOf2(size);
            _transferBuffer = CreateTransferBuffer(SDL_GPU_TRANSFERBUFFERUSAGE_UPLOAD, _transferBufferSize);

            _transferBufferOffset = 0;
        }

        if (size + _transferBufferOffset >= _transferBufferSize)
        {
            shouldCycle = true;
            _transferBufferOffset = 0;
        }

        offset = _transferBufferOffset;
        _transferBufferOffset += size;

        return _transferBuffer;
    }

    void RenderContext::CopyDataToTexture(SDL_GPUTexture* texture, void* data, const Vec2u& pos, const Sizeu& size, SDL_GPUTextureFormat format)
    {
        // multiply by 4 as RGBA8 is the only supported texture format right now
        // todo calculate the size of the pixel format once more pixelformats are supported
        u32 dataSize = size.Width * size.Height * 4;

        u32 offset;
        bool shouldCycle;
        SDL_GPUTransferBuffer* transBuffer = GetUploadBuffer(dataSize, offset, shouldCycle);

        CGE_TRACE("Transferring {}KiB of data to texture {} (offset: {}, cycle: {})",
            dataSize / 1024, reinterpret_cast<usize>(texture), offset, shouldCycle);

        void* mapped = SDL_MapGPUTransferBuffer(Device, transBuffer, shouldCycle);
        CGE_SDL_CHECK(mapped, "Map buffer");
        std::memcpy(static_cast<u8*>(mapped) + offset, data, dataSize);
        SDL_UnmapGPUTransferBuffer(Device, _transferBuffer);

        SDL_GPUCommandBuffer* cb = SDL_AcquireGPUCommandBuffer(Device);
        CGE_SDL_CHECK(cb, "Acquire command buffer");

        SDL_GPUCopyPass* pass = SDL_BeginGPUCopyPass(cb);
        CGE_SDL_CHECK(pass, "Begin copy pass");

        SDL_GPUTextureTransferInfo src
        {
            .transfer_buffer = transBuffer,
            .offset = offset,
            .pixels_per_row = size.Width,
            .rows_per_layer = size.Height
        };

        SDL_GPUTextureRegion dest
        {
            .texture = texture,
            .mip_level = 0,
            .layer = 0,
            .x = pos.X,
            .y = pos.Y,
            .z = 0,
            .w = size.Width,
            .h = size.Height,
            .d = 1
        };

        SDL_UploadToGPUTexture(pass, &src, &dest, false);

        SDL_EndGPUCopyPass(pass);
        CGE_SDL_CHECK(SDL_SubmitGPUCommandBuffer(cb), "Submit command buffer");
    }
}

File: src/Crimson.Graphics/RenderContext.cs
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using Crimson.Core;
using Crimson.Graphics.Utils;
using Crimson.Math;
using piko.SDL3;
using piko.SDL3.ShaderCross;

namespace Crimson.Graphics;

internal sealed class RenderContext : IDisposable
{
    /// <summary>
    /// 32MiB initial transfer buffer size.
    /// </summary>
    private const uint InitialTranferBufferSize = 32 * 1024 * 1024;

    private SDL.GPUTransferBuffer _transferBuffer;
    private uint _transferBufferSize;
    private uint _transferBufferOffset;

    public readonly SDL.Window Window;
    public readonly SDL.GPUDevice Device;

    public readonly HashSet<SDL.GPUTexture> MipmapQueue;

    public RenderContext(SDL.Window window)
    {
        Window = window;

        uint createProps = SDL.CreateProperties();

        if (OperatingSystem.IsWindows()) // enable dx12 on windows
            SDL.SetBooleanProperty(createProps, SDL.Prop.GpuDeviceCreateShadersDxilBoolean, true);
        else if (OperatingSystem.IsMacOS()) // enable metal on macos
            SDL.SetBooleanProperty(createProps, SDL.Prop.GpuDeviceCreateShadersMslBoolean, true);
        else // enable vulkan on everything else
            SDL.SetBooleanProperty(createProps, SDL.Prop.GpuDeviceCreateShadersSpirvBoolean, true);

#if DEBUG
        SDL.SetBooleanProperty(createProps, SDL.Prop.GpuDeviceCreateDebugmodeBoolean, true);
        SDL.SetBooleanProperty(createProps, SDL.Prop.GpuDeviceCreateVerboseBoolean, true);
#endif

        Logger.Trace("Creating device.");
        Device = SDL.CreateGPUDeviceWithProperties(createProps).Check("Create device");
        SDL.DestroyProperties(createProps);

        uint deviceProps = SDL.GetGPUDeviceProperties(Device);
        Logger.Info($"Backend: {SDL.GetGPUDeviceDriver(Device)}");
        Logger.Info($"Device: {SDL.GetStringProperty(deviceProps, SDL.Prop.GpuDeviceNameString, "unknown")}");
        Logger.Info($"Driver: {SDL.GetStringProperty(deviceProps, SDL.Prop.GpuDeviceDriverInfoString, "unknown")}");
        SDL.DestroyProperties(deviceProps);

        Logger.Trace("Claiming window for device.");
        SDL.ClaimWindowForGPUDevice(Device, Window).Check("Claim window for device");

        MipmapQueue = [];

        _transferBufferSize = InitialTranferBufferSize;
        _transferBufferOffset = 0;
        _transferBuffer = CreateTransferBuffer(SDL.GPUTransferBufferUsage.Upload, _transferBufferSize);
    }

    public unsafe SDL.GPUShader CreateShader(SDL.GPUShaderStage stage, string shader)
    {
        SDL.GPUShaderFormat format = SDL.GetGPUShaderFormats(Device);
        // sometimes multiple formats can be returned, this ensures that only one format
        // is set at a time as CreateGPUShader only allows for one format
        if ((format & SDL.GPUShaderFormat.Spirv) != 0)
            format = SDL.GPUShaderFormat.Spirv;
        else if ((format & SDL.GPUShaderFormat.Dxil) != 0)
            format = SDL.GPUShaderFormat.Dxil;
        else if ((format & SDL.GPUShaderFormat.Msl) != 0)
            format = SDL.GPUShaderFormat.Msl;
        else
            throw new PlatformNotSupportedException($"Unsupported shader format(s) {format}");

        string resourceName = $"Crimson.Graphics.{shader.Replace('/', '.')}.spv";
        byte[] spirv = Resource.Load(resourceName, Assembly.GetExecutingAssembly());

        ReadOnlySpan<byte> entryPoint = stage switch
        {
            SDL.GPUShaderStage.Vertex => "VSMain"u8,
            SDL.GPUShaderStage.Fragment => "PSMain"u8,
            _ => throw new ArgumentOutOfRangeException(nameof(stage), stage, null)
        };

        SDLShaderCross.GraphicsShaderMetadata* metadata;

        fixed (byte* pSpirv = spirv)
            metadata = SDLShaderCross.ReflectGraphicsSPIRV(pSpirv, (nuint) spirv.Length, 0);
        SDLUtils.Check((nint) metadata, "Reflect SPIRV");
        SDLShaderCross.GraphicsShaderResourceInfo resources = metadata->ResourceInfo;

        try
        {
            // no need to run shadercross if we're already on spirv
            // shadercross internally just passes it directly to CreateGPUShader and doesn't do anything,
            // but it saves an extra method call. meh yknow it makes no difference but whatever
            if (format == SDL.GPUShaderFormat.Spirv)
            {
                fixed (byte* pSpirv = spirv)
                fixed (byte* pEntryPoint = entryPoint)
                {
                    SDL.GPUShaderCreateInfo shaderInfo = new()
                    {
                        Code = pSpirv,
                        CodeSize = (nuint) spirv.Length,
                        Entrypoint = (sbyte*) pEntryPoint,
                        Stage = stage,
                        Format = format,
                        NumSamplers = resources.NumSamplers,
                        NumUniformBuffers = resources.NumUniformBuffers,
                        NumStorageTextures = resources.NumStorageTextures,
                        NumStorageBuffers = resources.NumStorageTextures
                    };

                    Logger.Trace($"Creating shader \"{shader}\"");
                    return SDL.CreateGPUShader(Device, &shaderInfo).Check("Create shader");
                }
            }

            fixed (byte* pSpirv = spirv)
            fixed (byte* pEntryPoint = entryPoint)
            {
                SDLShaderCross.SPIRVInfo spirvInfo = new()
                {
                    Bytecode = pSpirv,
                    BytecodeSize = (nuint) spirv.Length,
                    Entrypoint = (sbyte*) pEntryPoint,
                    ShaderStage = (SDLShaderCross.ShaderStage) stage
                };

                Logger.Trace($"Creating shader \"{shader}\"");
                return SDLShaderCross.CompileGraphicsShaderFromSPIRV(Device, &spirvInfo, &resources, 0)
                    .Check("Compile shader from spirv");
            }
        }
        finally
        {
            NativeMemory.Free(metadata);
        }
    }
    
    public unsafe SDL.GPUBuffer CreateBuffer(SDL.GPUBufferUsageFlags usage, uint size)
    {
        SDL.GPUBufferCreateInfo bufferInfo = new()
        {
            Usage = usage,
            Size = size
        };

        Logger.Trace($"Creating {size / 1024}KiB {usage} buffer.");
        return SDL.CreateGPUBuffer(Device, &bufferInfo).Check("Create buffer");
    }

    public unsafe SDL.GPUTransferBuffer CreateTransferBuffer(SDL.GPUTransferBufferUsage usage, uint size)
    {
        SDL.GPUTransferBufferCreateInfo bufferInfo = new()
        {
            Usage = usage,
            Size = size
        };

        Logger.Trace($"Creating {size / 1024}KiB transfer buffer.");
        return SDL.CreateGPUTransferBuffer(Device, &bufferInfo).Check("Create transfer buffer");
    }

    public SDL.GPUTransferBuffer GetUploadBuffer(uint size, out uint offset, out bool shouldCycle)
    {
        if (size >= _transferBufferSize)
        {
            Logger.Trace(
                $"Requested upload size ({size / 1024}KiB) is larger than the transfer buffer size ({_transferBufferSize / 1024}KiB). It will be resized.");
            // ensure the buffer can hold the data, then round to the next power of 2 so there's some wiggle room.
            _transferBufferSize = BitUtils.RoundToNextPowerOf2(size);
            _transferBufferOffset = 0;
            SDL.ReleaseGPUTransferBuffer(Device, _transferBuffer);
            _transferBuffer = CreateTransferBuffer(SDL.GPUTransferBufferUsage.Upload, _transferBufferSize);
        }

        shouldCycle = false;
        if (_transferBufferOffset + size >= _transferBufferSize)
        {
            shouldCycle = true;
            _transferBufferOffset = 0;
        }

        offset = _transferBufferOffset;
        _transferBufferOffset += size;

        return _transferBuffer;
    }

    // todo Vec2<uint> for X and Y
    public unsafe void CopyDataToTexture(SDL.GPUTexture texture, void* data, uint x, uint y, Size<uint> size,
        PixelFormat format)
    {
        uint dataSize = size.Width * size.Height * format.BytesPerPixel;

        SDL.GPUTransferBuffer transBuffer = GetUploadBuffer(dataSize, out uint offset, out bool cycle);
        Logger.Trace($"Uploading {dataSize / 1024}KiB data to texture {texture.Handle} (Offset: {offset}, Cycle: {cycle})");

        nint mapped = SDL.MapGPUTransferBuffer(Device, transBuffer, cycle).Check("Map transfer buffer");
        Unsafe.CopyBlock((byte*) mapped + offset, data, dataSize);
        SDL.UnmapGPUTransferBuffer(Device, transBuffer);

        SDL.GPUCommandBuffer cb = SDL.AcquireGPUCommandBuffer(Device).Check("Acquire command buffer");
        SDL.GPUCopyPass pass = SDL.BeginGPUCopyPass(cb).Check("Begin copy pass");

        SDL.GPUTextureTransferInfo src = new()
        {
            TransferBuffer = transBuffer,
            Offset = offset,
            PixelsPerRow = size.Width,
            RowsPerLayer = size.Height
        };

        SDL.GPUTextureRegion dest = new()
        {
            Texture = texture,
            X = x,
            Y = y,
            Z = 0,
            W = size.Width,
            H = size.Height,
            D = 1,
            MipLevel = 0,
            Layer = 0
        };

        SDL.UploadToGPUTexture(pass, &src, &dest, false);

        SDL.EndGPUCopyPass(pass);
        SDL.SubmitGPUCommandBuffer(cb).Check("Submit command buffer");
    }

    public void Dispose()
    {
        SDL.ReleaseGPUTransferBuffer(Device, _transferBuffer);
        SDL.ReleaseWindowFromGPUDevice(Device, Window);
        SDL.DestroyGPUDevice(Device);
    }
}